### PR TITLE
forbid bad types in indexing

### PR DESCRIPTION
**Pull request by drdzyk**
Forbid bad types in indexing arrays, strings and tuples

### CODE REVIEW

File: compiler/pipes/final-check.cpp [hunk 1]
```diff
@@ -527,6 +543,46 @@ void FinalCheckPass::check_instanceof(VertexAdaptor<op_instanceof> instanceof_ve
...
+                        what_indexing.empty() ? "" : " ", key_type.as_human_readable()));
+}
+
+static void check_indexing_type(PrimitiveType indexed_type, const TypeData &key_type) noexcept {
+  if (key_type.get_real_ptype() == tp_bool) {
+    return;
+  }
+
+  if (indexed_type == tp_tuple || indexed_type == tp_string) {
+    static const std::vector<PrimitiveType> allowed_types{tp_int, tp_float, tp_mixed, tp_future};
```
Comment (unserialize) [inline compiler/pipes/final-check.cpp:559]: I suppose, that for strings and tuples, we'd better allow only tp_int and tp_mixed: $s[1.0] seems unreasonable

File: compiler/pipes/final-check.cpp [hunk 2]
```diff
@@ -527,6 +543,46 @@ void FinalCheckPass::check_instanceof(VertexAdaptor<op_instanceof> instanceof_ve
...
+    check_indexing_violation("int, float and future", allowed_types, ptype_name(indexed_type), key_type);
+  } else {
+    static const std::vector<PrimitiveType> allowed_types{tp_int, tp_float, tp_string, tp_mixed, tp_future};
```
Comment (unserialize) [inline compiler/pipes/final-check.cpp:562]: Is tp_float reasonable for arrays indexing?

File: compiler/pipes/final-check.cpp (hunk 1 above)
Comment (drdzyk) [reply compiler/pipes/final-check.cpp]: Done

File: compiler/pipes/final-check.cpp (hunk 2 above)
Comment (drdzyk) [reply compiler/pipes/final-check.cpp]: To me not, but there are bunch of compilation issues in production. After discussion we decided that better leave as is.

File: compiler/pipes/final-check.cpp [hunk 3]
```diff
@@ -509,6 +515,16 @@ VertexPtr FinalCheckPass::on_enter_vertex(VertexPtr vertex) {
     check_instanceof(vertex.try_as<op_instanceof>());
   }
 
+  if (vertex->type() == op_index) {
+    const auto index_vertex = vertex.as<op_index>();
+    check_indexing(index_vertex->array(), index_vertex->key());
```
Comment (unserialize) [inline compiler/pipes/final-check.cpp:520]: Don't you want to use .try_as? Instread of checking type and casting:
```
if (const auto v_index = vertex.try_as<op_index>()) {
  check_indexing(v_index->array(), v_index->key());
} 
// and below
```

Comment (drdzyk) [reply compiler/pipes/final-check.cpp]: Done

**Review by DrDet — APPROVED** [review]

**Review by unserialize — APPROVED** [review]

**Pull request merged**